include/quic/parser.hpp

```diff
@@ -0,0 +1,378 @@
+#include <oxenc/bt.h>
+
+#include "stream.hpp"
+#include "utils.hpp"
+
+namespace oxen::quic
+{
+    inline auto bp_cat = oxen::log::Cat("bparser");
+
+    using time_point = std::chrono::steady_clock::time_point;
+
+    // timeout is used for sent requests awaiting responses
+    const std::chrono::seconds TIMEOUT{10};
+
+    // request sizes
+    inline constexpr long long MAX_REQ_LEN = 10_M;
+
+    // Application error
+    inline constexpr uint64_t BPARSER_EXCEPTION = (1ULL << 60) + 69;
+
+    struct message
+    {
+        std::string data;
+        std::string_view req_type;
+        std::string_view req_id;
+        std::string_view endpoint;
+        std::string_view req_body;
+
+        bool error{false};
+
+        message(std::string req, bool is_error = false) : data{std::move(req)}, error{is_error}
+        {
+            oxenc::bt_list_consumer btlc(data);
+
+            req_type = btlc.consume_string_view();
+            req_id = btlc.consume_string_view();
+
+            if (req_type == "Q" || req_type == "C")
+                endpoint = btlc.consume_string_view();
+
+            req_body = btlc.consume_string_view();
+        }
+
+        std::string rid() { return std::string{req_id}; }
+        std::string_view view() { return {data}; }
+    };
+
+    struct sent_request
+    {
+        // parsed request data
+        std::string data;
+        std::string req_id;
+
+        // total length of the request; is at the beginning of the request
+        size_t total_len;
+
+        std::chrono::steady_clock::time_point req_time;
+        std::chrono::steady_clock::time_point timeout;
+
+        bool is_empty() const { return data.empty() && total_len == 0; }
+
+        explicit sent_request(std::string d, std::string rid) : req_id{std::move(rid)}
+        {
+            total_len = d.length();
+            data += std::to_string(total_len);
+            data += ':';
+            data.reserve(data.length() + total_len);
+            data.append(d);
+
+            req_time = get_time();
+            timeout = req_time + TIMEOUT;
+        }
+
+        message to_message() { return {data}; }
+
+        std::string_view view() { return {data}; }
+        std::string&& payload() { return std::move(data); }
```

Comment:
```suggestion
        std::string&& payload() && { return std::move(data); }
```
so that you have to explicitly use `std::move(req).payload()` but just `req.payload()` wouldn't be allowed to destroy the data.